src/Options.dfy

```diff
@@ -0,0 +1,42 @@
+// RUN: %dafny "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+/*********************************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+**********************************************************************************************/
+
+module Options {
+  
+  datatype Option<T> = None | Some(value: T) {
+    function method to_result(): Result<T> 
+    {
+      match this
+      case Some(v) => Success(v)
+      case None() => Failure("Option is None")
+    }
+    
+    function method unwrap_or(default: T): T 
+    {
+      match this
+      case Some(v) => v
+      case None => default
+    }
+  }
+  
+  datatype Result<T> = | Success(value: T) | Failure(error: string) {
```

Comment:
In fact it can implement the entire "failure-compatible type" API; see https://dafny-lang.github.io/dafny//DafnyRef/DafnyRef#1971-failure-compatible-types.

Reply:
I'd also prefer this use a second type parameter for the error result rather than `string`.